Provide a simple program to inject arbitrary messages in the Hydra network

abailly-iohk opened the pull request:
This PR is the outcome of my experiments implementing a network injector to "attack" a Hydra cluster and DoS a Head which lead to catastrophic demo on July 27th.  It provides a `hydra-net` executable that can connect to any node and inject `ReqSn` messages. Of course, because the message is now signed it does not lead to the Head being stalled anymore, unless the signing key of one of the Hydra head participants are compromised.

---

<!-- Consider each and tick it off one way or the other -->
* [x] CHANGELOG updated or not needed
* [x] Documentation updated or not needed
* [x] Haddocks updated or not needed
* [x] No new TODOs introduced or explained herafter

abailly committed:
Scaffold executable to inject network messages

abailly committed:
Extract network server and client machinery to top-level functions

abailly committed:
First version of network injector

abailly committed:
Use lower-level function to send message to node

abailly committed:
Simplify injector client

abailly committed:
Expose network ports in docker demo

abailly committed:
Use connectToPeers instead of lower level actualConnect

abailly committed:
Revert "Use connectToPeers instead of lower level actualConnect"

This reverts commit 7f415e817f5a1661846250bdf948c4a45c562531.

abailly committed:
Add party to the ReqSn message

abailly committed:
Wrap message in heartbeat

Otherwise it does not get deserialised at the other end

abailly committed:
Adapt message sent to new authenticated format

abailly committed:
Use 'unstable' docker image

abailly committed:
Read a vk to impersonate some other party

abailly committed:
Revert most changes to Ouroboros module

We don't import much from it anymore as the amount of reuse is not so
large and it amounts to boilerplate machinery from the Ouroboros
framework.

abailly committed:
Improve handling of logs in hydra-net

abailly committed:
Add some documentation

v0d1ch commented on demo/docker-compose.yaml:
```diff
@@ -22,8 +22,8 @@ services:
-    # NOTE: Make sure to use the same image in ./seed-devnet.sh
-    image: ghcr.io/input-output-hk/hydra-node:0.11.0
+   # NOTE: Make sure to use the same image in ./seed-devnet.sh
+    image: ghcr.io/input-output-hk/hydra-node:unstable
```
We would overwrite these hydra-node versions upon each release. I guess you needed to use the latest on master for the demo and also it should be ok to keep it as is here since 0.12 will update it.

v0d1ch reviewed (approved):
BAM!

v0d1ch committed:
Remove stm and filepath dependencies

abailly committed:
Reinstate stable versions on docker-compose demo file

ch1bo commented on hydra-node/exe/hydra-net/Main.hs:
```diff
@@ -0,0 +1,205 @@
+{-# LANGUAGE TypeApplications #-}
+{-# OPTIONS_GHC -Wno-orphans #-}
```
I think this is not needed?

ch1bo committed:
Die on error instead of calling 'error'

This is showing a less "cryptic" error and can still be located easily
in the code base.

ch1bo committed:
Create hydra-net specific options parsers

This will improve help messages on the command line.

ch1bo reviewed (approved):
Was already a great PR. Improved the help messages myself a bit and definitely can approve this.

ch1bo commented:
Could run the executable against the demo setup, but obviously `MessageDropped` :)

ch1bo commented:
Merging this in spite of https://github.com/input-output-hk/hydra/actions/runs/5788084058/job/15686670468#step:4:131 as this is unrelated.
